wfe2/wfe.go

```diff
@@ -1548,7 +1549,23 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
 		requesterAccount = acct
 	}
 
+	requestedChain := 0
 	serial := request.URL.Path
+
+	// An alternate chain may be requested with the request path {serial}/{chain}, where chain
+	// is a number - an index into the slice of chains for the issuer. If a specific chain is
+	// not requested, then it defaults to zero - the default certificate chain for the issuer.
+	serialAndChain := strings.SplitN(serial, "/", 2)
+	if len(serialAndChain) == 2 {
+		idx, err := strconv.Atoi(serialAndChain[1])
+		if err != nil {
+			wfe.sendError(response, logEvent, probs.NotFound("Certificate not found"), nil)
```

Comment:
I think this makes more sense as `probs.Malformed("Chain ID must be an integer")` or something.

Reply:
I reasoned that being more specific would be exposing the chain ID as a public API, when the client should treat the URL as opaque without interpreting it. Happy to make the change anyway if you want, just a thought.

Reply:
I think `NotFound` is right here. Here's an example of where we give a generic NotFound when a challenge ID doesn't exist as part of an authorization:

https://github.com/letsencrypt/boulder/blob/e92c5c66ea89517c8d3422b09c3a355c53430a40/wfe2/wfe.go#L1048-L1052

Reply:
I think this is a closer example of what we already do: https://github.com/letsencrypt/boulder/blob/e92c5c66ea89517c8d3422b09c3a355c53430a40/wfe2/wfe.go#L1028-L1032

The challenge ID isn't an integer/we don't check its formatting, so not found makes more sense there.

Reply:
Just to clarify my thinking here, if the ID _could_ be a valid slice index then not found makes sense (i.e. >= 0, which is what happens later), but if it's an invalid argument then malformed makes more sense.

Reply:
>but if it's an invalid argument then malformed makes more sense.

In that case, should the `core.ValidSerial` check also be changed to use malformed?

Reply:
Eh, I'd probably do that if we were introducing new code now, but I think it's outside the scope of this PR.

Reply:
@jsha since you already approved the latest set, did you still want these last set of changes suggested by Roland?

```diff
diff --git a/wfe2/wfe.go b/wfe2/wfe.go
index f305b793..09171040 100644
--- a/wfe2/wfe.go
+++ b/wfe2/wfe.go
@@ -1559,7 +1559,8 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
        if len(serialAndChain) == 2 {
                idx, err := strconv.Atoi(serialAndChain[1])
                if err != nil || idx < 0 {
-                       wfe.sendError(response, logEvent, probs.NotFound("Certificate not found"), nil)
+                       wfe.sendError(response, logEvent, probs.Malformed("Chain ID must be a non-negative integer"),
+                               fmt.Errorf("certificate chain id provided was not valid: %s", serialAndChain[1]))
                        return
                }
                serial = serialAndChain[0]
@@ -1651,7 +1652,7 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
                // If the requested chain is outside the bounds of the available chains,
                // then it is an error by the client - not found.
                if requestedChain < 0 || requestedChain >= len(availableChains) {
-                       wfe.sendError(response, logEvent, probs.NotFound("Certificate not found"), nil)
+                       wfe.sendError(response, logEvent, probs.NotFound("Unknown issuance chain"), nil)
                        return
                }
```

Reply:
Thanks for the notice - I missed that you hadn't yet addressed Roland's request for changes. Please do add that in.